### PR TITLE
Feature: Align brackets

### LINKED ISSUE #453: Feature request: more flexible record whitespace
Issue created from [fantomas-ui](https://jindraivanek.gitlab.io/fantomas-ui/#?fantomas=preview-trivia&code=C4TwDgpgBAYg9nKBeAUFdUDeaO5gSwgBsATKALinwDtgdd0DiSAmCq2lehp0lgFn7sadbnkJ8AzMM5iMvVgFYZohvImsAbCu4BfIA&config=N4IgkgdgJgphAuBlADgQwMYwHIFcC2IAXACwA0IACqgOYwDqAllPABZEAcADORQE4wBnGLwBuMAKLQA8gDMAMgwgwiM1ABsh5RDDwN0AezX6IAQXiSoshUpXrNIFBhgAhGDP38Tvavjjwi8Lw4MFpomK7u-ADChsYBQSEOYTAmMvDCMXh4qPHBoU6p6bzaugZGELmJkLAIUhAAKrwAnoysthqJAEowHrC8UshwACIw6GqovKjwDHGEqh35mF76ONAjagy6RZVagXrwALL6sJUAvkA)

G-Research's internal coding guidelines specify that if a record spans more than one line, it should be split one entry per line, and the braces must be symmetrical. The original source below conforms to G-Research's coding standards; Fantomas's output does not. Could we please have a config option for this? I know that the official F# style guidelines say that this layout is "unusual", but it is the completely standard layout for us.

#### Code

```fsharp
type Foo =
    {
        Field : int
        Field2 : int
        Field244 : int
        Field3 : int
    }
```
            

#### Result

```fsharp
type Foo =
    { Field : int
      Field2 : int
      Field244 : int
      Field3 : int }

```
            
#### Options

Fantomas Next - 3.0.0-7/20/2019

| Name | Value |
| ---- | ----- |
| `IndentSpaceNum` | `4` |
| **`PageWidth`** | **`80`** |
| `PreserveEndOfLine` | `false` |
| `SemicolonAtEndOfLine` | `false` |
| `SpaceBeforeArgument` | `true` |
| **`SpaceBeforeColon`** | **`true`** |
| `SpaceAfterComma` | `true` |
| `SpaceAfterSemicolon` | `true` |
| `IndentOnTryWith` | `false` |
| `ReorderOpenDeclaration` | `false` |
| `SpaceAroundDelimiter` | `true` |
| **`StrictMode`** | **`true`** |

### PR DESCRIPTION
**Pull request by nojaf**
Alternative style of formatting records, arrays and lists.
Would resolve #453 
Might need some more usecases and be tested in the wild.

### CODE REVIEW

File: docs/Documentation.md [hunk 1]
```diff
@@ -286,6 +286,65 @@ That said, most of the preferences are very simple.
 But they demonstrate the flexibility of Fantomas on a set of configurations. 
 More preferences will be added depending on use cases.
 
+##### "AlignBrackets"
+
+Alternative style to format records, arrays and lists.
+If enabled, this will align the brackets (`{`, `{|`, `[`, `[|`) and (`}`, `|}`, `]`, `|]`) on the same line.
```
Comment (auduchinok) [inline docs/Documentation.md]: It seems it should've been `on separate lines` here?

File: docs/Documentation.md [hunk 2]
```diff
@@ -306,6 +365,7 @@ A default configuration file would look like
   "SpaceAroundDelimiter":true ,
   "KeepNewlineAfter":false,
   "MaxIfThenElseShortWidth":40,
+  "AlignBrackets": false,
```
Comment (auduchinok) [inline docs/Documentation.md]: Could it be a more descriptive name maybe? `BracketsOnSeparateLines` or something?

File: docs/Documentation.md (hunk 1 above)
Comment (nojaf) [reply docs/Documentation.md]: Separate lines isn't always the case. For example: https://github.com/fsprojects/fantomas/pull/706/files#diff-2bb5a182cc5d09abf8ec12ca99f7d645R93

That is why I felt like align brackets was a description that covered all scenarios.
`BracketsOnSameColumn` might be an alternative, I'm definitely open to suggestions.

Comment (auduchinok) [reply docs/Documentation.md]: I see, thanks. The on the same line part looks misleading, though. I'm trying to come up with a better name.

Comment (auduchinok) [reply docs/Documentation.md]: Maybe something like `MultilineBlockBracketsOnSeparateLines`?

Comment (nojaf) [reply docs/Documentation.md]: Maybe `MultilineBlockBracketsOnSameColumn` ?
The `OnSeparateLines` part might create confusing for
```fsharp
{ new System.Object() with
            member x.ToString() = "F#"
}
```
You could get the impression that `{` should be on a separate line.
However, this is by design not the case for object expressions and record updates.

File: src/Fantomas.Tests/MultilineBlockBracketsOnSameColumnRecordTests.fs [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+open FsUnit
+open Fantomas.Tests.TestHelper
+
+let config = ({ config with
+                    MultilineBlockBracketsOnSameColumn = true
+                    SpaceBeforeColon = true
+                    SpaceBeforeSemicolon = true })
+
+[<Test>]
+let ``single member record stays on oneline`` () =
```
Comment (JackMatusiewicz) [inline src/Fantomas.Tests/MultilineBlockBracketsOnSameColumnRecordTests.fs]: I think there should be a space between one and line

Comment (nojaf) [reply src/Fantomas.Tests/MultilineBlockBracketsOnSameColumnRecordTests.fs]: Indeed, thanks.

**Comment by nojaf** [comment]
Still WIP, requires #721 to be merged first.

**Comment by nojaf** [comment]
Closing in favour of #746

**Pull request closed without merging**